### ISSUE TITLE
Rustc fails to suggest an easy fix; instead provides a confusing error message

### ISSUE THREAD

**Issue by digitalnomd** [rust-lang/rust#72124]
The following code produces an unexpected error message:

```rust
test.split_whitespace().next().ok_or_else(|| {
    format_err!("Couldn't split the test string");
})?;
```

I expected the compiler to give me a warning like "`ok_or_else` needs to return the trait `std::error::Error`. Instead you are returning `()`. Suggestion: remove the `;`.

Instead, I received the following error message which was hard to parse:
```
 |     })?;
  |       ^ the trait `std::error::Error` is not implemented for `()`
  |
  = note: required because of the requirements on the impl of `failure::Fail` for `()`
  = note: required because of the requirements on the impl of `std::convert::From<()>` for `failure::Error`
  = note: required by `std::convert::From::from`
```

The rust compiler version that I am using is `1.43.1`. This behavior also exists within the beta and nightly versions. I also created an example using the rust-playground here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=33e42842b0d6385ea5704e323c113e7f.

Thanks again!
